### ISSUE TITLE
Address mapping and return flow routing

### ISSUE THREAD

**Issue by DavidSchinazi** [ietf-wg-masque/draft-ietf-masque-connect-udp#5]
Issue created by @martinthomson on [2020-04-16](https://github.com/DavidSchinazi/masque-drafts/issues/25):

The design here would appear to permit the use of 'Address and Port-Dependent Mapping' as defined in [Section 4.1 of RFC 4787](https://tools.ietf.org/html/rfc4787#section-4.1).  However, that document strongly recommends that an 'Endpoint-Independent Mapping' is used so as to make UNSAF methods work.  This document should define those rules.

Related #4.

**Comment by DavidSchinazi** [comment]
Comment from me on 2020-04-17:

The current text explicitly only allows "Address and Port-Dependent Mapping". It attempted to define those rules clearly in [section 6](https://tools.ietf.org/html/draft-schinazi-masque-connect-udp-00#section-6). UNSAF is currently considered out of scope and not supported. I'm not sure whether we want to increase the scope and add that complexity.

**Comment by DavidSchinazi** [comment]
@martinthomson can you elaborate on what you would like to see happen here? Ideally in the form of a PR?

**Comment by gloinul** [comment]
So at a minimal functionality wise is that a MASQUE client should be able to request to open additional destinations from the same proxy source address (ip+UDP port) so that the client can open for UDP flows that are return routed to the proxy for a client from an address the client didn't expect but later learned. This is also necessary to get some of the ICE scenarios with STUN/TURN servers and direct connections to work. This is a step between a general listening socket, and the strict bi-direcitonal 5-tuple connection that is currently defined.

**Comment by gloinul** [comment]
I will note that this issue describes an intermedia step between a server port (#35 ) and the basic 5-tuple. And I think it is fundamental aspects of the protocol design to enable this.

**Comment by martinthomson** [comment]
I think that there are two distinct use cases that might need to be treated as separate things.

1. The client wants to initiate a flow toward a specific UDP endpoint.  For that you don't need to understand what IP and port the proxy is allocating and that address does not need to be accessible from other (remote) endpoints.  This would work for QUIC; it might work in some very limited peer-to-peer and media cases, but I consider that not especially likely.
2. The client wants a UDP endpoint of its own to use.  For that, the request is really that the proxy allocate an IP+port for exclusive use of the client.  In this case, the proxy should probably communicate what it thinks the address is to the client (though the client might seek the opinion of a STUN server regarding a reflexive address).  This would be necessary for most RTP and peer-to-peer scenarios.

My suggestion here is to look into an extension, perhaps via a header field, for the second case and only promise the former without that extension being present.

**Comment by DavidSchinazi** [comment]
I agree with @martinthomson here, this would make for a useful extension.

**Comment by DavidSchinazi** [comment]
Based on discussion, it sounds like this will be better served by an extension. So I'm closing this issue with no action. Please reopen/comment if you disagree.
